File: rust/src/cgroup.rs
//! cgroup v2 operations: create, set limits, kill, cleanup.

use std::collections::HashMap;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

const CGROUP_BASE: &str = "/sys/fs/cgroup/nitrobox";

/// Check if cgroup v2 is available on this system.
#[must_use]
pub fn cgroup_v2_available() -> bool {
    Path::new("/sys/fs/cgroup/cgroup.controllers").exists()
}

/// Create a cgroup for the given sandbox name. Returns the cgroup path.
pub fn create_cgroup(name: &str) -> io::Result<PathBuf> {
    let path = PathBuf::from(CGROUP_BASE).join(name);
    fs::create_dir_all(&path)?;
    Ok(path)
}

/// Enable required controllers on the parent cgroup.
pub fn enable_controllers(cgroup_path: &Path, limits: &HashMap<String, String>) -> io::Result<()> {
    let Some(parent) = cgroup_path.parent() else {
        return Ok(());
    };
    let subtree_ctl = parent.join("cgroup.subtree_control");
    if !subtree_ctl.exists() {
        return Ok(());
    }

    let controller_map = [
        ("cpu_max", "cpu"),
        ("cpu_shares", "cpu"),
        ("memory_max", "memory"),
        ("memory_swap", "memory"),
        ("pids_max", "pids"),
        ("io_max", "io"),
        ("cpuset_cpus", "cpuset"),
    ];

    for (key, ctrl) in &controller_map {
        if limits.contains_key(*key) {
            let _ = fs::write(&subtree_ctl, format!("+{ctrl}"));
        }
    }
    Ok(())
}

/// Convert Docker CPU shares (2-262144) to cgroup v2 weight (1-10000).
#[must_use]
pub fn convert_cpu_shares(shares: u64) -> u64 {
    if shares == 0 {
        return 100; // default
    }
    // Docker shares range: 2-262144, default 1024
    // cgroup v2 weight range: 1-10000, default 100
    ((shares.saturating_sub(2)) * 9999 / 262_142 + 1).clamp(1, 10000)
}

/// Apply resource limits to a cgroup.
pub fn apply_limits(cgroup_path: &Path, limits: &HashMap<String, String>) -> io::Result<()> {
    let limit_files = [
        ("cpu_max", "cpu.max"),
        ("memory_max", "memory.max"),
        ("pids_max", "pids.max"),
        ("io_max", "io.max"),
        ("cpuset_cpus", "cpuset.cpus"),
        ("cpu_shares", "cpu.weight"),
        ("memory_swap", "memory.swap.max"),
    ];

    for (key, filename) in &limit_files {
        if let Some(value) = limits.get(*key) {
            let write_value = if *key == "cpu_shares" {
                match value.parse::<u64>() {
                    Ok(shares) => convert_cpu_shares(shares).to_string(),
                    Err(_) => continue,
                }
            } else {
                value.clone()
            };

            let file_path = cgroup_path.join(filename);
            match fs::write(&file_path, &write_value) {
                Ok(()) => log::debug!("cgroup {filename} = {write_value}"),
                Err(e) => log::warn!("Failed to set cgroup {filename}: {e}"),
            }
        }
    }
    Ok(())
}

/// Move a process into a cgroup.
pub fn add_process(cgroup_path: &Path, pid: u32) -> io::Result<()> {
    fs::write(cgroup_path.join("cgroup.procs"), pid.to_string())
}

/// Kill all processes in a cgroup and remove it.
pub fn cleanup_cgroup(cgroup_path: &Path) -> io::Result<()> {
    if !cgroup_path.exists() {
        return Ok(());
    }

    // Kill via cgroup.kill (kernel 5.14+)
    let kill_file = cgroup_path.join("cgroup.kill");
    if kill_file.exists() {
        let _ = fs::write(&kill_file, "1");
    }

    // SIGKILL any remaining processes
    let procs_file = cgroup_path.join("cgroup.procs");
    if procs_file.exists() {
        if let Ok(contents) = fs::read_to_string(&procs_file) {
            for pid_str in contents.split_whitespace() {
                if let Ok(pid) = pid_str.parse::<i32>() {
                    let _ = nix::sys::signal::kill(
                        nix::unistd::Pid::from_raw(pid),
                        nix::sys::signal::Signal::SIGKILL,
                    );
                }
            }
        }
    }

    // Remove the cgroup directory
    match fs::remove_dir(cgroup_path) {
        Ok(()) => {}
        Err(e) => log::debug!("cgroup cleanup (non-fatal): {e}"),
    }
    Ok(())
}

File: go/internal/cgroup/cgroup.go
// Package cgroup provides cgroup v2 operations.
package cgroup

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"golang.org/x/sys/unix"
)

const cgroupBase = "/sys/fs/cgroup/nitrobox"

// V2Available returns true if cgroup v2 is available.
func V2Available() bool {
	_, err := os.Stat("/sys/fs/cgroup/cgroup.controllers")
	return err == nil
}

// Create creates a cgroup and returns its path.
func Create(name string) (string, error) {
	path := filepath.Join(cgroupBase, name)
	if err := os.MkdirAll(path, 0o755); err != nil {
		return "", err
	}
	return path, nil
}

// EnableControllers enables required controllers on the parent cgroup.
func EnableControllers(cgroupPath string, limits map[string]string) error {
	parent := filepath.Dir(cgroupPath)
	subtreeCtl := filepath.Join(parent, "cgroup.subtree_control")
	if _, err := os.Stat(subtreeCtl); err != nil {
		return nil
	}

	controllerMap := map[string]string{
		"cpu_max":     "cpu",
		"cpu_shares":  "cpu",
		"memory_max":  "memory",
		"memory_high": "memory",
		"memory_swap": "memory",
		"pids_max":    "pids",
		"io_max":      "io",
		"cpuset_cpus": "cpuset",
		"cpuset_mems": "cpuset",
	}

	for key, ctrl := range controllerMap {
		if _, ok := limits[key]; ok {
			_ = os.WriteFile(subtreeCtl, []byte("+"+ctrl), 0o644)
		}
	}
	return nil
}

// ConvertCPUShares converts Docker CPU shares (2-262144) to cgroup v2 weight (1-10000).
func ConvertCPUShares(shares uint64) uint64 {
	if shares == 0 {
		return 100
	}
	w := (shares - 2) * 9999 / 262142
	w += 1
	if w < 1 {
		return 1
	}
	if w > 10000 {
		return 10000
	}
	return w
}

// ApplyLimits applies resource limits to a cgroup.
func ApplyLimits(cgroupPath string, limits map[string]string) error {
	limitFiles := map[string]string{
		"cpu_max":     "cpu.max",
		"memory_max":  "memory.max",
		"memory_high": "memory.high",
		"pids_max":    "pids.max",
		"io_max":      "io.max",
		"cpuset_cpus": "cpuset.cpus",
		"cpuset_mems": "cpuset.mems",
		"cpu_shares":  "cpu.weight",
		"memory_swap": "memory.swap.max",
	}

	for key, filename := range limitFiles {
		value, ok := limits[key]
		if !ok {
			continue
		}

		writeValue := value
		if key == "cpu_shares" {
			shares, err := strconv.ParseUint(value, 10, 64)
			if err != nil {
				continue
			}
			writeValue = strconv.FormatUint(ConvertCPUShares(shares), 10)
		}

		filePath := filepath.Join(cgroupPath, filename)
		if err := os.WriteFile(filePath, []byte(writeValue), 0o644); err != nil {
			// Non-fatal, log and continue
			fmt.Fprintf(os.Stderr, "warning: failed to set cgroup %s: %v\n", filename, err)
		}
	}
	return nil
}

// AddProcess moves a process into a cgroup.
func AddProcess(cgroupPath string, pid uint32) error {
	return os.WriteFile(filepath.Join(cgroupPath, "cgroup.procs"), []byte(strconv.FormatUint(uint64(pid), 10)), 0o644)
}

// Cleanup kills all processes in a cgroup and removes it.
func Cleanup(cgroupPath string) error {
	if _, err := os.Stat(cgroupPath); err != nil {
		return nil
	}

	// Kill via cgroup.kill (kernel 5.14+)
	killFile := filepath.Join(cgroupPath, "cgroup.kill")
	if _, err := os.Stat(killFile); err == nil {
		_ = os.WriteFile(killFile, []byte("1"), 0o644)
	}

	// SIGKILL remaining processes
	procsFile := filepath.Join(cgroupPath, "cgroup.procs")
	data, err := os.ReadFile(procsFile)
	if err == nil {
		for _, pidStr := range strings.Fields(string(data)) {
			pid, err := strconv.Atoi(pidStr)
			if err == nil {
				_ = unix.Kill(pid, unix.SIGKILL)
			}
		}
	}

	// Retry rmdir
	for i := 0; i < 20; i++ {
		err := os.Remove(cgroupPath)
		if err == nil {
			return nil
		}
		if i < 19 {
			time.Sleep(10 * time.Millisecond)
		}
	}
	return nil
}
